### PR TITLE
add new parsers - pmd, tfsec

### PR DESCRIPTION
**Pull request by jis0324**
Added new parsers - pmd, tfsec

### CODE REVIEW

**Comment by damiencarol** [comment]
@jis0324 why closing this PR? your PR was good. There is some code to add but nothing hard.

**Comment by jis0324** [comment]
Hi, @damiencarol 
Actually, I was working with the forked repository.
Then I sent pull request to here, I think it was my fault.
Btw, Could you allow my request?

**Comment by jis0324** [comment]
If possible, I think I can open this request again.
Btw, How did you find my requests? actually, I deleted it soon since I requested it wrong.
lol..

**Review by damiencarol — CHANGES_REQUESTED** [review]
- split the 2 parsers in 2 PR
 - don't need to modify dojo/fixtures/test_type.json anymore
 - please get ride of meta attribute like `__author__`
 - address flake8 errors

**Comment by jis0324** [comment]
About : `split the 2 parsers in 2 PR`
Should I create two PR?
Or should I commit twice times in same PR?

**Comment by jis0324** [comment]
I seem to address your requirements.

**Comment by damiencarol** [comment]
@jis0324 better to create 2 PR for 2 different parsers. I'm sorry to ask you more work :D

File: dojo/tools/pmd/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__author__ = 'jis0324'
```
Comment (damiencarol) [inline dojo/tools/pmd/__init__.py:1]:
```suggestion
```
This is useless. We track authorship with our VCS.

File: dojo/tools/pmd/parser.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            elif row["Priority"] == "4":
+                priority = "High"
+            elif row["Priority"] == "3":
+                priority = "Medium"
+            elif row["Priority"] == "2":
+                priority = "Low"
+            elif row["Priority"] == "1":
+                priority = "Info"
+            else:
+                priority = row["Priority"]
```
Comment (damiencarol) [inline dojo/tools/pmd/parser.py:44]: priority will be a number like "1" here.
```suggestion
                priority = "Info"
```

File: dojo/tools/pmd/parser.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+            description = "Description: {}\n".format(row['Description'].strip())
+            description += "Rule set: {}\n".format(row["Rule set"].strip())
+            description += "Rule: {}\n".format(row["Rule"].strip())
+            finding.description = description
+            finding.line = row["Line"]
+            finding.file_path = row["File"]
+            finding.component_name = row["Package"]
+
+            if finding is not None:
```
Comment (damiencarol) [inline dojo/tools/pmd/parser.py:55]: this test is useless because you assign an object at 31

File: dojo/tools/pmd/parser.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                if finding.title is None:
```
Comment (damiencarol) [inline dojo/tools/pmd/parser.py:56]: this test is useless because you assign a value at 32

File: dojo/tools/pmd/parser.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                    finding.title = ""
+                if finding.description is None:
```
Comment (damiencarol) [inline dojo/tools/pmd/parser.py:58]: this test is useless because you assign a value at 50

File: dojo/tools/pmd/parser.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                    finding.description = ""
+
+                key = hashlib.md5((finding.title + '|' + finding.description).encode("utf-8")).hexdigest()
```
Comment (damiencarol) [inline dojo/tools/pmd/parser.py:61]: Don't use MD5, it's flawed
```suggestion
                key = hashlib.sha256((finding.title + '|' + finding.description).encode("utf-8")).hexdigest()
```

File: dojo/tools/tfsec/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+__author__ = 'jis0324'
```
Comment (damiencarol) [inline dojo/tools/tfsec/__init__.py:1]:
```suggestion
```
This is useless. We track authorship with our VCS.

File: dojo/tools/tfsec/parser.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+        for row in csvarray:
+            finding = Finding(test=test)
+            if row["severity"] == "5":
+                priority = "Critical"
+            elif row["severity"] == "4":
+                priority = "High"
+            elif row["severity"] == "3":
+                priority = "Medium"
+            elif row["severity"] == "WARNING":
```
Comment (damiencarol) [inline dojo/tools/tfsec/parser.py:38]: This warning seems out of context.

File: dojo/tools/tfsec/parser.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            # finding.line = row["start_line"] + "~" + row["end_line"]
+            finding.file_path = row["file"]
+
+            if "for more information" in row["link"]:
+                try:
+                    finding.url = row["link"].rsplit("for more information", 1)[0].strip().split(" ")[1].strip()
+                except:
+                    finding.url = row["link"].rsplit("for more information", 1)[0].strip()
+
+            if finding is not None:
```
Comment (damiencarol) [inline dojo/tools/tfsec/parser.py:59]: same comment as for the other parser.

File: dojo/unittests/tools/test_pmd_parser.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class TestPMDParser(TestCase):
+
+    def test_parse_file_with_no_vuln_has_no_findings(self):
+        testfile = open("dojo/unittests/scans/pmd/pmd_no_vuln.csv")
+        parser = PmdParser()
+        findings = parser.get_findings(testfile, Test())
+        self.assertEqual(0, len(findings))
+
```
Comment (damiencarol) [inline dojo/unittests/tools/test_pmd_parser.py:13]: Add more test on attributes like:
```suggestion
        self.assertEqual("Medium", findings[0].severity)

```

File: dojo/unittests/tools/test_tfsec_parser.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class TestTfsecParser(TestCase):
+
+    def test_parse_file_with_no_vuln_has_no_findings(self):
+        testfile = open("dojo/unittests/scans/tfsec/tfsec_no_vuln.csv")
+        parser = TfsecParser()
+        findings = parser.get_findings(testfile, Test())
+        self.assertEqual(0, len(findings))
+
```
Comment (damiencarol) [inline dojo/unittests/tools/test_tfsec_parser.py:13]: Add more tests on attributes

**Review by damiencarol — CHANGES_REQUESTED** [review]

**Comment by Maffooch** [comment]
@jis0324 Whenever you do split this into two PR's please base them from them `dev` branch and then submit the PR against the `dev` branch so you don't get any funky conflicts or extra commits. Thanks!

**Comment by jis0324** [comment]
> @jis0324 better to create 2 PR for 2 different parsers. I'm sorry to ask you more work :D

No, It's okay, I am not familiar with GitHub's works.
This is my first PR. lol...

**Comment by jis0324** [comment]
Btw, I don't know how I can re-create PRs.
Could you give me any help or idea?

**Comment by damiencarol** [comment]
@jis0324 you can create more than one branch in your fork. you can create on PR for each branch.
but if it's more easy, you can finish this one by keeping only one parser and after create a new one with the second parser.

**Comment by jis0324** [comment]
@damiencarol
Make sense what you mean.
So Should I close this PR?
Then I will creat new PR to dev branch. right?

**Comment by jis0324** [comment]
Which branch should I start from?
I mean which branch I will clone and update.

Clone `dev` branch, and then Request PR to `dev` branch.
Is right?

**Comment by damiencarol** [comment]
yes. exactly

**Pull request closed without merging**